### PR TITLE
feat: prefer take to filter for very sparse masks

### LINKED ISSUE #1248: When applying RowMask to chunk choose between filter and take
very sparse masks are likely faster with take than filter

### PR DESCRIPTION
**Pull request by danking**
Fixes https://github.com/spiraldb/vortex/issues/1248

It seems to me that if we have fewer than one-in-1024 values there is no chance for SIMD vectorization to make the mask as fast as a take.

Mild improvement.

```
group                       develop                 layouts                 take-one-in-1024        take-one-in-128         take-one-in-16         always-take
-----                       -------                 -------                 ----------------        ---------------         --------------         -----------
vortex-local-fs             1.00   381.3±10.78µs    34.91    13.3±0.13ms    29.49    11.2±0.11ms    29.63    11.3±0.09ms    29.75    11.3±0.13ms   29.71    11.3±0.12ms
vortex-tokio-local-disk     1.00   349.5±12.64µs    29.46    10.3±0.14ms    24.23     8.5±0.15ms    24.73     8.6±0.09ms    24.80     8.7±0.12ms   24.85     8.7±0.18ms
```

### CODE REVIEW

**Comment by robert3005** [comment]
But always take is still faster than filter which is suspicious?

File: vortex-serde/src/layouts/read/mask.rs [hunk 1]
```diff
@@ -158,12 +158,20 @@ impl RowMask {
...
+            take(sliced, indices).map(Some)
+        } else {
+            let mask = self.to_mask_array()?;
+            filter(sliced, mask).map(Some)
+        }
+    }
 
-        filter(sliced, predicate).map(Some)
+    pub fn to_indices_array(&self) -> VortexResult<Array> {
+        Ok(PrimitiveArray::from_vec(self.values.to_vec(), Validity::NonNullable).into_array())
```
Comment (robert3005) [inline vortex-serde/src/layouts/read/mask.rs:173]: I guess we actually need roaring arrays for this to avoid the copy here

**Comment by danking** [comment]
The benchmark does not explore very many densities:
```
const INDICES: [u64; 6] = [10, 11, 12, 13, 100_000, 3_000_000];
```

File: vortex-serde/src/layouts/read/mask.rs (hunk 1 above)
Comment (danking) [reply vortex-serde/src/layouts/read/mask.rs:173]: Oh let me try using roaring int

Comment (danking) [reply vortex-serde/src/layouts/read/mask.rs:173]: Erm that won't do us much good, the take functions convert to primitive anyway.

Comment (robert3005) [reply vortex-serde/src/layouts/read/mask.rs:173]: yes sorry you do need a bit more specialization across the board. Also roaring int array has serialization cost. This is not a simple fix but potentially perf

**Review by robert3005 — APPROVED** [review]
Can you extract the threshold as a constant in the file?

**Comment by robert3005** [comment]
fix #1248

**Comment by danking** [comment]
Done.

**Pull request merged**